### PR TITLE
logging: VASSERT and environment variable control

### LINKED ISSUE #153: Add argument to `GALOIS_LOG_ASSERT` to print message
Current version does not take any additional arguments, but having an optional message can be useful to give some context.

https://github.com/KatanaGraph/katana/pull/149#discussion_r450398539

### PR DESCRIPTION
**Pull request by witchel**
Wishes can come true. 

Closes #153

### CODE REVIEW

File: libsupport/include/galois/Logging.h [hunk 1]
```diff
@@ -63,9 +65,20 @@ void Log(LogLevel level, F fmt_string, Args&&... args) {
 template <typename F, typename... Args>
 void LogLine(LogLevel level, const char* file_name, int line_no, F fmt_string,
              Args&&... args) {
-  std::string s         = fmt::format(fmt_string, std::forward<Args>(args)...);
-  std::string with_line = fmt::format("{}:{}: {}", file_name, line_no, s);
-  internal::LogString(level, with_line);
+  int env_log_level = static_cast<int32_t>(LogLevel::Debug);
+  GetEnv("KATANA_LOG_LEVEL", &env_log_level);
```
Comment (ddn0) [inline libsupport/include/galois/Logging.h:69]: my preference would be to stick with `GALOIS_` until we do a single global replace to `KATANA_`

File: libsupport/include/galois/Logging.h [hunk 2]
```diff
@@ -63,9 +65,20 @@ void Log(LogLevel level, F fmt_string, Args&&... args) {
...
+  std::string s = fmt::format(fmt_string, std::forward<Args>(args)...);
+  std::string with_line;
+  // Supress printing of File and line information with an environment variable
+  if (!GetEnv("KATANA_SUPPRESS_FILE_LINE")) {
```
Comment (ddn0) [inline libsupport/include/galois/Logging.h:73]: If we shorten the paths in `file_name` would it be okay again to just print file and line information?

I'd to reduce the number of runtime options if possible.

File: libsupport/include/galois/Logging.h [hunk 3]
```diff
@@ -63,9 +65,20 @@ void Log(LogLevel level, F fmt_string, Args&&... args) {
...
+    with_line = fmt::format("{}:{}: {}", file_name, line_no, s);
+  } else {
+    with_line = s;
+  }
+  // Only log KATANA_LOG_LEVEL and above (default, log everything)
+  if (static_cast<int32_t>(level) >= env_log_level) {
```
Comment (ddn0) [inline libsupport/include/galois/Logging.h:79]: if we go with `_LOG_LEVEL` we should probably just drop `GALOIS_LOG_VERBOSE` in https://github.com/KatanaGraph/katana/blob/master/libsupport/src/Logging.cpp#L32

Small bonus points for moving the implementation of `LOG_LEVEL` down to `Logging.cpp` too: it reduces code size slightly and keeps the template creep down.

File: libsupport/include/galois/Logging.h (hunk 1 above)
Comment (witchel) [reply libsupport/include/galois/Logging.h:69]: Oops, my bad.

File: libsupport/include/galois/Logging.h (hunk 3 above)
Comment (witchel) [reply libsupport/include/galois/Logging.h:79]: GALOIS_LOG_LEVEL isn't terribly type safe, but I like the control.

File: libsupport/include/galois/Logging.h (hunk 2 above)
Comment (witchel) [reply libsupport/include/galois/Logging.h:73]: Can you explain that to me?  Because my instincts are the opposite--provide tools to get the info you need in a format that is easy to process.  But certainly controlling log level with an environment variable is more important.

Comment (ddn0) [reply libsupport/include/galois/Logging.h]: I guess there are two concerns:

1.  "reduce the number of runtime options". Avoiding more things to support and document even if internally. Also, simplifying onsite debugging: "Can you send over the logs after enabling Foo and Bar and recompiling with DebugFoo? ... Oh you have set Foo already but at level Green? That explains why we were missing some messages."

2. "provide tools to get the info you need in a format that is easy to process". I agree with this. Though, my go-to to do this would be to pick a uniform "container" format (CSV, JSON, etc.) and then pretty print things in a final presentation layer. I suppose there is a question about who is doing the processing (e.g., a human reading a terminal, a script reading the logs). FWIW, I tend to think of the primary consumer of log output as another program as it is not particularly easy to continually monitor n programs by hand.

Long story short, I was hoping that if we shortened the file name output, we could enter a world where we don't need to have a suppress file name option.

Comment (witchel) [reply libsupport/include/galois/Logging.h]: I get it.  I think you are right, I can filter.  I removed that code.
It would be nice if the strings "DEBUG: ", "VERBOSE: ", "WARNING: ", and "ERROR: " were absent from any of our messages because that would make parsing the log output easier.
I moved the LOG_LEVEL change into Logging.cpp.
